### DISCUSSION TITLE
How to use HomeBrew to create universal libraries

### DISCUSSION THREAD

**Discussion by cbfiddle** [Homebrew/discussions#4647] [Everyday usage] [answered]
### Output of `brew config`

```shell
NA
```

### Output of `brew doctor`

```shell
NA
```

### Description of issue

I have started building macOS universal applications for which I need universal libraries.
That means I need both x86 and arm versions of libraries that I currently install using HomeBrew.
My understanding is that I need two installations of HomeBrew, one on an x86 machine and one on an arm machine, to bulid the x86 and arm library versions. Is that the best option?
(I've read about tricky ways of downloading libraries for the "other" architecture, but I don't want to use tricky solutions.)
I assume that running /usr/local/bin/brew on an arm machine creates arm libraries, just more slowly. Is that correct?

**Comment by SMillerDev** [comment]
Either way it's going to be tricky because this is not something Homebrew is designed to support.

> I assume that running /usr/local/bin/brew on an arm machine creates arm libraries, just more slowly. Is that correct?

No, `/usr/local/bin/brew` is an x86 install. `/opt/Homebrew/bin/brew` is the default arm install

**Reply by cbfiddle** [reply]
> Either way it's going to be tricky because this is not something Homebrew is designed to support.

Can you give me some idea of what the tricky part is? I don't understand how running HomeBrew on two machines to get libraries for different architectures is tricky.

> No, /usr/local/bin/brew is an x86 install.

You are saying it would download x86 libraries? That sounds tricky, but if it worked it would be more convenient.

**Reply by SMillerDev** [reply]
> Can you give me some idea of what the tricky part is? I don't understand how running HomeBrew on two machines to get libraries for different architectures is tricky.

Getting the libraries to work outside of Homebrew, this is not at all a usecase we optimise for.

**Reply by cbfiddle** [reply]
> Getting the libraries to work outside of Homebrew, this is not at all a usecase we optimise for.

You're talking about the shared library ID and dependencies?

**Reply by cbfiddle** [reply]
> No, /usr/local/bin/brew is an x86 install.

This seems odd, then:

```
$ /usr/local/bin/brew doctor
Warning: Your Homebrew's prefix is not /opt/homebrew.
```

```
/usr/local/bin/brew upgrade
Error: Cannot install in Homebrew on ARM processor in Intel default prefix (/usr/local)!
Please create a new installation in /opt/homebrew using one of the
"Alternative Installs" from:
  https://docs.brew.sh/Installation
You can migrate your previously installed formula list with:
  brew bundle dump
```

**Reply by alebcay** [reply]
Did you install Homebrew in a terminal running under Rosetta? Under Rosetta, which emulates x86_64, Homebrew will still install to `/usr/local/bin`. This allows users to simultaneously have an installation at `/opt/homebrew` (for arm64) and at `/usr/local/bin` (for x86_64, only when running under Rosetta).

What may have happened here is that you installed Homebrew while under Rosetta, hence it being installed to the `/usr/local/bin` prefix by default, and are now attempting to run Homebrew while not under Rosetta, hence the error.

**Reply by cbfiddle** [reply]
Actually, I was not reporting a problem, but trying to clarify a previous comment that implied the ability to download x86 binaries on an arm64 system using /usr/local/bin/brew (which apparently was incorrect?).

I'm not sure what you mean by running Homebrew while "under Rosetta" or "not under Rosetta". If /usr/local/bin/brew is an x86 executable, it will only run "under Rosetta", won't it?

How it got there is by being copied from an x86 system, either by me or by Migration Assistant.

**Reply by alebcay** [reply]
`/usr/local/bin/brew` is a shell script. I imagine there is a check inside that checks `uname` or other aspects of the environment to see if it is being invoked in the native arm64 environment or inside an emulated x86_64 environment, and that check is what is showing the error message you see.

I will try to answer your original question in a top-level reply, I understand you're not trying to report an issue here.

**Comment by alebcay** [comment, marked as answer]
To expand upon Sean's response, Homebrew is not currently designed to distribute formulae as universal binaries (or libraries). What this means is that if you run `brew` on an Intel x86_64 machine, you will get a copy of the libraries for x86_64; if you run `brew` on an arm64 machine, you will get a copy of the libraries for arm64.

You could install one copy of Homebrew on both an x86_64 and arm64 machine, copy the shared libraries, and then use `lipo` to join them into a universal library. You could also do the same on a single arm64 machine by installing two copies of Homebrew, one invoked under Rosetta 2 (i.e. `arch -x86_64 /usr/local/bin/brew install ...`).

Regardless of whether you try to install two copies of Homebrew on one machine or two, like Sean mentioned this is probably going to cause some other problems. Programs and libraries installed via Homebrew are generally not designed to be re-distributed outside of Homebrew. For instance, the rpath of the dylib is set to the specific prefix where Homebrew installed the library and the location of any dependencies it may require are also set explicitly to point to locations where Homebrew installed these dependencies.

In short, Homebrew performs some actions on shared libraries/executables distributed via Homebrew to ensure that things "just work" in Homebrew. These actions may cause these shared libraries/executables to stop working if one attempts to use them outside of Homebrew.

I would suggest using a different approach for obtaining the libraries that you need to distribute with your application. There are other package managers/ecosystems out there that are designed with redistributable build artifacts in mind, such as CMake, vcpkg, conan, or build2, depending on your needs. These other ecosystems may also provide the ability to cross-compile/build a single universal copy at once since they offer control over compilation options.

**Reply by cbfiddle** [reply]
Thank you.

It does appear that `arch -x86_64 /usr/local/bin/brew` works, although it might be safer to run brew x86 on an x86 machine and sync a copy of /usr/local to my arm64 machine.

I'm not sure that using a different package manager would make things easier. I already have code that replaces the Shared Library ID with @rpath... and uses lipo to glue together x86 and arm64 libraries. I did not see any instructions for obtaining universal libraries using the package managers you mention.

**Reply by alebcay** [reply]
You may be right, I don't see a particularly easy way to obtain fat libraries from the other ecosystems either (conan, for instance, is single-config, i.e. 1 binary/library per architecture). It sounds like you may have to do some `lipo`-ing either way, but it may be easier to obtain files for the other architecture using the other methods, since Homebrew always assumes host architecture = target architecture to be installed (except with the Rosetta trick, as mentioned).

It sounds like you've figured out how you might go about using Homebrew-provided libraries (the approach of copying, and sounds like you have a process to make it work). I just advise against it because the use case you describe is something we never test, nor is it considered a very big issue on our radar if it were not to work (i.e. we really only have the bandwidth to care about it working if you `brew install` it; not if you `brew install` it, copy it into your application, distribute the application, and the end user of the application proceeds to have issues with it).

**Reply by cbfiddle** [reply]
I appreciate your advice. It is not clear to me what a possible failure mode might be using HomeBrew libraries or why libraries obtained from some other package manager might not have the same possible failure mode. Maybe I will find out...

**Reply by alebcay** [reply]
Some examples off the top of my head are that [we sometimes apply patches that aren't accepted upstream to fix things (a fix from our view, maybe a kludge from others' perspective)](https://github.com/Homebrew/homebrew-core/blob/1aebdef0550f438e654952101480e52cf8b18b10/Formula/vsftpd.rb#L28-L41) (we try to minimize this as much as possible but there are some cases where it can't be avoided). We don't check if these patches break things outside of use as installed via Homebrew.

In other cases, [we may have to patch out some functionality for licensing/legal reasons](https://github.com/Homebrew/homebrew-core/blob/1aebdef0550f438e654952101480e52cf8b18b10/Formula/p7zip.rb#L45-L71).

Homebrew (at least the core repository) also has no notion of optional dependencies so [some libraries just default to having most if not all features enabled, which often results in more dependencies](https://github.com/Homebrew/homebrew-core/blob/1aebdef0550f438e654952101480e52cf8b18b10/Formula/ffmpeg.rb#L88-L127).

These may not be concerns depending on the exact libraries you're looking to use, of course. I guess the biggest caution is still just "we don't use it regularly ourselves for such a purpose - nor is it a case we intentionally support, so we don't even know fully what _could_ break".

**Reply by cbfiddle** [reply]
Got it. I think I would notice those examples locally (before I exported the libraries in an application) if they caused problems. There could _in theory_ be changes that have impact only in the enviroment where the application is used, but aside from linking issues, I would not expect any in the libraries that I am using, which are purely computational (image rendering and FFT). If we were talking about mariaDB, things might be different.
